graphql/src/execution/execution.rs

```diff
@@ -15,87 +15,149 @@ use crate::values::coercion;
-    pub introspection_schema: &'a s::Document,
+    pub schema: Arc<Schema>,
 
     /// The query to execute.
     pub document: &'a q::Document,
 
     /// The resolver to use.
-    pub resolver: Arc<R1>,
-
-    /// The introspection resolver to use.
-    pub introspection_resolver: Arc<R2>,
+    pub resolver: Arc<R>,
 
     /// The current field stack (e.g. allUsers > friends > name).
     pub fields: Vec<&'a q::Field>,
 
-    /// Whether or not we're executing an introspection query
-    pub introspecting: bool,
-
     /// Variable values.
     pub variable_values: Arc<HashMap<q::Name, q::Value>>,
 
     /// Time at which the query times out.
     pub deadline: Option<Instant>,
 }
 
-impl<'a, R1, R2> ExecutionContext<'a, R1, R2>
+impl<'a, R> ExecutionContext<'a, R>
 where
-    R1: Resolver,
-    R2: Resolver,
+    R: Resolver,
 {
     /// Creates a derived context for a new field (added to the top of the field stack).
-    pub fn for_field(&mut self, field: &'a q::Field) -> Self {
+    pub fn for_field(&self, field: &'a q::Field) -> Self {
         let mut ctx = self.clone();
         ctx.fields.push(field);
         ctx
     }
+
+    pub fn as_introspection_context(&self) -> ExecutionContext<IntrospectionResolver> {
+        // Create an introspection type store and resolver
+        let introspection_schema = introspection_schema(self.schema.id.clone());
+        let introspection_resolver = IntrospectionResolver::new(&self.logger, &self.schema);
+
+        ExecutionContext {
+            logger: self.logger.clone(),
+            resolver: Arc::new(introspection_resolver),
+            schema: Arc::new(introspection_schema),
+            document: &self.document,
+            fields: vec![],
+            variable_values: self.variable_values.clone(),
+            deadline: self.deadline,
+        }
+    }
 }
 
 /// Executes the root selection set of a query.
-pub fn execute_root_selection_set<'a, R1, R2>(
-    ctx: ExecutionContext<'a, R1, R2>,
+pub fn execute_root_selection_set<'a, R>(
+    ctx: &ExecutionContext<'a, R>,
     selection_set: &'a q::SelectionSet,
     initial_value: &Option<q::Value>,
 ) -> Result<q::Value, Vec<QueryExecutionError>>
 where
-    R1: Resolver,
-    R2: Resolver,
+    R: Resolver,
 {
     // Obtain the root Query type and fail if there isn't one
     let query_type = match sast::get_root_query_type(&ctx.schema.document) {
         Some(t) => t,
         None => return Err(vec![QueryExecutionError::NoRootQueryObjectType]),
     };
 
+    // Split the toplevel fields into introspection fields and
+    // 'normal' data fields
+    let mut data_set = q::SelectionSet {
+        span: selection_set.span.clone(),
+        items: Vec::new(),
+    };
+    let mut intro_set = q::SelectionSet {
+        span: selection_set.span.clone(),
+        items: Vec::new(),
+    };
+
+    let ictx = ctx.as_introspection_context();
+    let introspection_query_type = sast::get_root_query_type(&ictx.schema.document).unwrap();
```

Comment:
By the spec, introspection queries should be executed against the same root type as the regular queries.

Reply:
I skimmed through the spec but I didn't find it to be very specific about this. I don't think it makes a difference here.

Reply:
In the spec introspection doesn't have a full schema, but is set of fields and types that are implicitly injected into any root query type. This does seem equivalent as long as the query types are the same wrt to directives and arguments (we currently have none of those anyways), and this change will help with sql based execution, so I'm not saying we shouldn't do this, but just to be mindful of this detail.

Reply:
It's also worth mentioning that this ultimately does things the same way we used to do them, just that the switch to the introspection schema is more explicit and at the root level.